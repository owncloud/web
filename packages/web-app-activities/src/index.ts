import translations from '../l10n/translations.json'
import { useGettext } from 'vue3-gettext'
import { computed } from 'vue'
import { AppMenuItemExtension, defineWebApplication, Extension } from '@ownclouders/web-pkg'
import { urlJoin } from '@ownclouders/web-client'
import { RouteRecordRaw } from 'vue-router'
import { APPID } from './appid'

export default defineWebApplication({
  setup() {
    const { $gettext } = useGettext()

    const appInfo = {
      name: $gettext('Activities'),
      id: APPID,
      icon: 'pulse',
      color: '#887ef1'
    }

    const routes: RouteRecordRaw[] = [
      {
        path: '/',
        name: 'root',
        component: () => import('./LayoutContainer.vue'),
        redirect: urlJoin(appInfo.id, 'list'),
        meta: {
          authContext: 'user'
        },
        children: [
          {
            path: 'list',
            name: 'list',
            component: () => import('./views/App.vue'),
            meta: {
              authContext: 'user',
              title: $gettext('Activities')
            }
          }
        ]
      }
    ]

    const menuItemExtension: AppMenuItemExtension = {
      id: `app.${appInfo.id}.menuItem`,
      type: 'appMenuItem',
      label: () => appInfo.name,
      color: appInfo.color,
      icon: appInfo.icon,
      priority: 30,
      path: urlJoin(appInfo.id)
    }
    const extensions = computed(() => {
      const result: Extension[] = []

      result.push(menuItemExtension)

      return result
    })

    return {
      appInfo,
      routes,
      translations,
      extensions
    }
  }
})
