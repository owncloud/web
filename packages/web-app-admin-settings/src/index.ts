import translations from '../l10n/translations.json'
import General from './views/General.vue'
import Users from './views/Users.vue'
import Groups from './views/Groups.vue'
import Spaces from './views/Spaces.vue'
import { Ability, urlJoin } from '@ownclouders/web-client'
import {
  ApplicationInformation,
  AppMenuItemExtension,
  AppNavigationItem,
  defineWebApplication,
  useAbility,
  useUserStore
} from '@ownclouders/web-pkg'
import { RouteRecordRaw } from 'vue-router'
import { computed } from 'vue'

// just a dummy function to trick gettext tools
function $gettext(msg: string) {
  return msg
}

const appId = 'admin-settings'

function adminAcrGuard(to, from, next) {
  if (typeof window === 'undefined') {
    next()
    return
  }

  // Avoid re-triggering when returning from step-up or when a request is already in-flight for this route
  const hasStepupParam = !!to.query.stepup
  // const hasGlobalStepupFlag = new URLSearchParams(window.location.search).has('stepupAcr')
  const stepupKey = 'oc.stepup:admin-settings'
  const alreadyRequested = sessionStorage.getItem(stepupKey) === '1'

  // diagnostics
  try {
    console.debug('[adminAcrGuard] enter', {
      to: to.fullPath,
      hasStepupParam,
      // hasGlobalStepupFlag,
      alreadyRequested,
      locationSearch: window.location.search
    })
  } catch (_) {}

  // if (!hasStepupParam && !hasGlobalStepupFlag && !alreadyRequested) {
  if (!hasStepupParam && !alreadyRequested) {
    const redirectUrl = to.fullPath + (to.fullPath.includes('?') ? '&' : '?') + 'stepup=1'
    try {
      sessionStorage.setItem(stepupKey, '1')
    } catch (_) {
      // ignore storage errors
    }
    try {
      console.info('[adminAcrGuard] initiating step-up', { redirectUrl })
    } catch (_) {}
    window.location.assign(`/?stepupAcr=advanced&redirectUrl=${encodeURIComponent(redirectUrl)}`)
    return
  }

  // window.location.assign(`/`)
  // return
  try {
    console.debug('[adminAcrGuard] pass-through')
  } catch (_) {}
  next()
}

function getAvailableRoute(ability: Ability) {
  if (ability.can('read-all', 'Setting')) {
    return { name: 'admin-settings-general' }
  }

  if (ability.can('read-all', 'Account')) {
    return { name: 'admin-settings-users' }
  }

  if (ability.can('read-all', 'Group')) {
    return { name: 'admin-settings-groups' }
  }

  if (ability.can('read-all', 'Drive')) {
    return { name: 'admin-settings-spaces' }
  }

  throw Error('Insufficient permissions')
}

export const routes = ({ $ability }: { $ability: Ability }): RouteRecordRaw[] => [
  {
    path: '/',
    redirect: () => {
      return { name: 'admin-settings-general' }
    }
  },
  {
    path: '/general',
    name: 'admin-settings-general',
    component: General,
    beforeEnter: adminAcrGuard,
    meta: {
      authContext: 'user',
      title: $gettext('General')
    }
  },
  {
    path: '/users',
    name: 'admin-settings-users',
    component: Users,
    beforeEnter: adminAcrGuard,
    meta: {
      authContext: 'user',
      title: $gettext('Users')
    }
  },
  {
    path: '/groups',
    name: 'admin-settings-groups',
    component: Groups,
    beforeEnter: adminAcrGuard,
    meta: {
      authContext: 'user',
      title: $gettext('Groups')
    }
  },
  {
    path: '/spaces',
    name: 'admin-settings-spaces',
    component: Spaces,
    beforeEnter: adminAcrGuard,
    meta: {
      authContext: 'user',
      title: $gettext('Spaces')
    }
  }
]

export const navItems = ({ $ability }: { $ability: Ability }): AppNavigationItem[] => [
  {
    name: $gettext('General'),
    icon: 'settings-4',
    route: {
      path: `/${appId}/general?`
    },
    isVisible: () => {
      return $ability.can('read-all', 'Setting')
    },
    priority: 10
  },
  {
    name: $gettext('Users'),
    icon: 'user',
    route: {
      path: `/${appId}/users?`
    },
    isVisible: () => {
      return $ability.can('read-all', 'Account')
    },
    priority: 20
  },
  {
    name: $gettext('Groups'),
    icon: 'group-2',
    route: {
      path: `/${appId}/groups?`
    },
    isVisible: () => {
      return $ability.can('read-all', 'Group')
    },
    priority: 30
  },
  {
    name: $gettext('Spaces'),
    icon: 'layout-grid',
    route: {
      path: `/${appId}/spaces?`
    },
    isVisible: () => {
      return $ability.can('read-all', 'Drive')
    },
    priority: 40
  }
]

export default defineWebApplication({
  setup() {
    const { can } = useAbility()
    const userStore = useUserStore()

    const appInfo: ApplicationInformation = {
      name: $gettext('Admin Settings'),
      id: appId,
      icon: 'settings-4',
      color: '#2b2b2b'
    }

    const menuItems = computed<AppMenuItemExtension[]>(() => {
      const items: AppMenuItemExtension[] = []

      const menuItemAvailable =
        userStore.user &&
        (can('read-all', 'Setting') ||
          can('read-all', 'Account') ||
          can('read-all', 'Group') ||
          can('read-all', 'Drive'))

      if (menuItemAvailable) {
        items.push({
          id: `app.${appInfo.id}.menuItem`,
          type: 'appMenuItem',
          label: () => appInfo.name,
          color: appInfo.color,
          icon: appInfo.icon,
          priority: 40,
          path: urlJoin(appInfo.id)
        })
      }

      return items
    })

    return {
      appInfo,
      routes,
      navItems,
      translations,
      extensions: menuItems
    }
  }
})
