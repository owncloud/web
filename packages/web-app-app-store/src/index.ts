import translations from '../l10n/translations.json'
import { useGettext } from 'vue3-gettext'
import { computed, unref } from 'vue'
import {
  AppMenuItemExtension,
  defineWebApplication,
  Extension,
  useAbility,
  useUserStore
} from '@ownclouders/web-pkg'
import { urlJoin } from '@ownclouders/web-client'
import { RouteRecordRaw } from 'vue-router'
import { useRepositoriesStore } from './piniaStores'
import { AppStoreConfigSchema, AppStoreRepository } from './types'
import { APPID } from './appid'

export default defineWebApplication({
  setup({ applicationConfig }) {
    const { $gettext } = useGettext()
    const { can } = useAbility()
    const userStore = useUserStore()
    const repositoryStore = useRepositoriesStore()

    const defaultRepositories: AppStoreRepository[] = [
      {
        name: 'awesome-ocis',
        url: 'https://raw.githubusercontent.com/owncloud/awesome-ocis/main/webApps/apps.json'
      }
    ]
    if (applicationConfig?.repositories) {
      const { repositories } = AppStoreConfigSchema.parse(applicationConfig)
      repositoryStore.setRepositories(repositories || defaultRepositories)
    } else {
      repositoryStore.setRepositories(defaultRepositories)
    }

    const appInfo = {
      name: $gettext('App Store'),
      id: APPID,
      icon: 'store',
      color: '#ff6961'
    }

    const hasPermission = computed(() => {
      // TODO: which permission(s) do we need to check here?
      return userStore.user && can('read-all', 'Setting')
    })

    const routes: RouteRecordRaw[] = [
      {
        path: '/',
        name: 'root',
        component: () => import('./LayoutContainer.vue'),
        redirect: urlJoin(appInfo.id, 'list'),
        beforeEnter: (to, from, next) => {
          if (!unref(hasPermission)) {
            return next({ path: '/' })
          }
          next()
        },
        meta: {
          authContext: 'user'
        },
        children: [
          {
            path: 'list',
            name: 'list',
            component: () => import('./views/AppList.vue'),
            meta: {
              authContext: 'user',
              title: $gettext('App Store')
            }
          },
          {
            path: 'app/:appId',
            name: 'details',
            component: () => import('./views/AppDetails.vue'),
            meta: {
              authContext: 'user',
              title: $gettext('App Details')
            }
          }
        ]
      }
    ]

    const menuItemExtension: AppMenuItemExtension = {
      id: `app.${appInfo.id}.menuItem`,
      type: 'appMenuItem',
      label: () => appInfo.name,
      color: appInfo.color,
      icon: appInfo.icon,
      priority: 30,
      path: urlJoin(appInfo.id)
    }
    const extensions = computed(() => {
      const result: Extension[] = []

      if (unref(hasPermission)) {
        result.push(menuItemExtension)
      }

      return result
    })

    return {
      appInfo,
      routes,
      translations,
      extensions
    }
  }
})
