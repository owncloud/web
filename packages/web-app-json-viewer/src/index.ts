import { AppWrapperRoute } from '@ownclouders/web-pkg'
import JsonViewer from './App.vue'

// just a dummy function to trick gettext tools
function $gettext(msg: string) {
  return msg
}

const routes = [
  {
    path: '/:driveAliasAndItem(.*)?',
    component: AppWrapperRoute(JsonViewer, {
      applicationId: 'json-viewer'
    }),
    name: 'json-viewer',
    meta: {
      authContext: 'hybrid',
      title: $gettext('JSON Viewer'),
      patchCleanPath: true
    }
  }
]

const appInfo = {
  name: $gettext('JSON Viewer'),
  id: 'json-viewer',
  icon: 'file-code',
  iconFillType: 'fill',
  extensions: [
    {
      extension: 'json',
      routeName: 'json-viewer'
    }
  ]
}

export default {
  appInfo,
  routes
}
