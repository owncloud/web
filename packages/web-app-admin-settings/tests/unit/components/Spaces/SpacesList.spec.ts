import SpacesList from '../../../../src/components/Spaces/SpacesList.vue'
import { defaultComponentMocks, defaultPlugins, mount, shallowMount } from 'web-test-helpers'
import { SortDir, eventBus, queryItemAsString } from '@ownclouders/web-pkg'
import { displayPositionedDropdown } from '@ownclouders/web-pkg'
import { SideBarEventTopics } from '@ownclouders/web-pkg'
import { nextTick } from 'vue'
import { useSpaceSettingsStore } from '../../../../src/composables'
import { mock } from 'vitest-mock-extended'
import { OcTable } from 'design-system/src/components'
import { GraphSharePermission, SpaceMember, SpaceResource } from '@ownclouders/web-client'

const spaceMocks = [
  mock<SpaceResource>({
    id: '1',
    name: '1 Some space',
    disabled: false,
    members: {
      '1': mock<SpaceMember>({
        permissions: [GraphSharePermission.deletePermissions],
        grantedTo: { user: { displayName: 'user1' } }
      }),
      '2': mock<SpaceMember>({
        permissions: [GraphSharePermission.deletePermissions],
        grantedTo: { user: { displayName: 'user2' } }
      }),
      '3': mock<SpaceMember>({
        permissions: [GraphSharePermission.deletePermissions],
        grantedTo: { user: { displayName: 'user3' } }
      })
    },
    spaceQuota: {
      total: 1073741824,
      used: 0,
      remaining: 1073741824
    }
  }),
  mock<SpaceResource>({
    id: '2',
    name: '2 Another space',
    disabled: true,
    members: {
      '1': mock<SpaceMember>({
        permissions: [GraphSharePermission.deletePermissions],
        grantedTo: { user: { displayName: 'user1' } }
      }),
      '2': mock<SpaceMember>({
        permissions: [GraphSharePermission.deletePermissions],
        grantedTo: { user: { displayName: 'user2' } }
      }),
      '3': mock<SpaceMember>({
        permissions: [GraphSharePermission.deletePermissions],
        grantedTo: { user: { displayName: 'user3' } }
      }),
      '4': mock<SpaceMember>({
        permissions: [],
        grantedTo: { user: { displayName: 'user4' } }
      }),
      '5': mock<SpaceMember>({
        permissions: [],
        grantedTo: { user: { displayName: 'user5' } }
      })
    },
    spaceQuota: {
      total: 2147483648,
      used: 524288000,
      remaining: 1610612736
    }
  })
]

const selectors = {
  ocTableStub: 'oc-table-stub'
}

vi.mock('@ownclouders/web-pkg', async (importOriginal) => ({
  ...(await importOriginal<any>()),
  displayPositionedDropdown: vi.fn(),
  queryItemAsString: vi.fn()
}))

describe('SpacesList', () => {
  it('should render all spaces in a table', () => {
    const { wrapper } = getWrapper({ spaces: spaceMocks })
    expect(wrapper.html()).toMatchSnapshot()
  })
  it.each(['name', 'members', 'totalQuota', 'usedQuota', 'remainingQuota', 'status'])(
    'sorts by property "%s"',
    async (prop) => {
      const { wrapper } = getWrapper({ mountType: shallowMount, spaces: spaceMocks })
      wrapper.vm.sortBy = prop
      await wrapper.vm.$nextTick()
      expect(
        (
          wrapper.findComponent<typeof OcTable>(selectors.ocTableStub).props()
            .data[0] as SpaceResource
        ).id
      ).toBe(spaceMocks[0].id)
      wrapper.vm.sortDir = SortDir.Desc
      await wrapper.vm.$nextTick()
      expect(
        (
          wrapper.findComponent<typeof OcTable>(selectors.ocTableStub).props()
            .data[0] as SpaceResource
        ).id
      ).toBe(spaceMocks[1].id)
    }
  )
  it('should set the sort parameters accordingly when calling "handleSort"', () => {
    const { wrapper } = getWrapper({ spaces: [spaceMocks[0]] })
    const sortBy = 'members'
    const sortDir = SortDir.Desc
    wrapper.vm.handleSort({ sortBy, sortDir })
    expect(wrapper.vm.sortBy).toEqual(sortBy)
    expect(wrapper.vm.sortDir).toEqual(sortDir)
  })
  it('shows only filtered spaces if filter applied', async () => {
    const { wrapper } = getWrapper({ spaces: spaceMocks })
    wrapper.vm.filterTerm = 'Another'
    await nextTick()
    expect(wrapper.vm.items).toEqual([spaceMocks[1]])
  })
  it('should show the context menu on right click', async () => {
    const spyDisplayPositionedDropdown = vi.mocked(displayPositionedDropdown)
    const { wrapper } = getWrapper({ spaces: spaceMocks })
    await wrapper.find(`[data-item-id="${spaceMocks[0].id}"]`).trigger('contextmenu')
    expect(spyDisplayPositionedDropdown).toHaveBeenCalledTimes(1)
  })
  it('should show the context menu on context menu button click', async () => {
    const spyDisplayPositionedDropdown = vi.mocked(displayPositionedDropdown)
    const { wrapper } = getWrapper({ spaces: spaceMocks })
    await wrapper.find('.spaces-table-btn-action-dropdown').trigger('click')
    expect(spyDisplayPositionedDropdown).toHaveBeenCalledTimes(1)
  })
  it('should show the space details on details button click', async () => {
    const eventBusSpy = vi.spyOn(eventBus, 'publish')
    const { wrapper } = getWrapper({ spaces: spaceMocks })
    await wrapper.find('.spaces-table-btn-details').trigger('click')
    expect(eventBusSpy).toHaveBeenCalledWith(SideBarEventTopics.open)
  })
  describe('toggle selection', () => {
    describe('selectSpaces method', () => {
      it('selects all spaces', () => {
        const spaces = [
          mock<SpaceResource>({ id: '1', name: 'Some Space' }),
          mock<SpaceResource>({ id: '2', name: 'Some other Space' })
        ]
        const { wrapper } = getWrapper({ mountType: shallowMount, spaces })
        wrapper.vm.selectSpaces(spaces)
        const { setSelectedSpaces } = useSpaceSettingsStore()
        expect(setSelectedSpaces).toHaveBeenCalledWith(spaces)
      })
    })
    describe('selectSpace ', () => {
      it('selects a space', () => {
        const spaces = [mock<SpaceResource>({ id: '1', name: 'Some Space' })]
        const { wrapper } = getWrapper({ mountType: shallowMount, spaces })
        wrapper.vm.selectSpace(spaces[0])
        const { addSelectedSpace } = useSpaceSettingsStore()
        expect(addSelectedSpace).toHaveBeenCalledWith(spaces[0])
      })
      it('de-selects a selected space', () => {
        const spaces = [mock<SpaceResource>({ id: '1', name: 'Some Space' })]
        const { wrapper } = getWrapper({ mountType: shallowMount, spaces, selectedSpaces: spaces })
        wrapper.vm.selectSpace(spaces[0])
        const { setSelectedSpaces } = useSpaceSettingsStore()
        expect(setSelectedSpaces).toHaveBeenCalledWith([])
      })
    })
    describe('unselectAllSpaces method', () => {
      it('de-selects all selected spaces', () => {
        const spaces = [mock<SpaceResource>({ id: '1', name: 'Some Space' })]
        const { wrapper } = getWrapper({ mountType: shallowMount, spaces })
        wrapper.vm.unselectAllSpaces()
        const { setSelectedSpaces } = useSpaceSettingsStore()
        expect(setSelectedSpaces).toHaveBeenCalledWith([])
      })
    })
  })
})

function getWrapper({
  mountType = mount,
  spaces = [],
  selectedSpaces = []
}: { mountType?: typeof mount; spaces?: SpaceResource[]; selectedSpaces?: SpaceResource[] } = {}) {
  vi.mocked(queryItemAsString).mockImplementationOnce(() => '1')
  vi.mocked(queryItemAsString).mockImplementationOnce(() => '100')
  const mocks = defaultComponentMocks()

  return {
    wrapper: mountType(SpacesList, {
      props: {
        headerPosition: 0
      },
      global: {
        plugins: [
          ...defaultPlugins({
            piniaOptions: {
              spaceSettingsStore: { spaces, selectedSpaces }
            }
          })
        ],
        mocks,
        provide: mocks,
        stubs: {
          OcCheckbox: true
        }
      }
    })
  }
}
