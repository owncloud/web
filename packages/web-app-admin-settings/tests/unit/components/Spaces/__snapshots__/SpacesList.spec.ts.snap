// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SpacesList > should render all spaces in a table 1`] = `
"<div id="space-list" headerposition="0">
  <div class="space-filters oc-flex oc-flex-right oc-flex-wrap oc-flex-bottom oc-mx-m oc-mb-m">
    <div class=""><label class="oc-label" for="spaces-filter">Search</label>
      <div class="oc-position-relative">
        <!--v-if--> <input id="spaces-filter" autocomplete="off" aria-invalid="false" class="oc-text-input oc-input oc-rounded" type="text" value="">
        <!--v-if-->
      </div>
      <!--v-if-->
      <!--v-if-->
    </div>
  </div>
  <table class="oc-table oc-table-hover oc-table-sticky has-item-context-menu spaces-table">
    <thead class="oc-thead">
      <tr tabindex="-1" class="oc-table-header-row">
        <th class="oc-table-cell oc-table-cell-align-left oc-table-cell-align-middle oc-table-cell-width-shrink oc-text-nowrap oc-th oc-table-header-cell oc-table-header-cell-select oc-pl-s" style="top: 0px;">
          <div><span class="oc-table-thead-content"><oc-checkbox-stub id="oc-checkbox-1" disabled="false" modelvalue="false" label="Select all spaces" labelhidden="false" hidelabel="true" size="large" outline="false" class="oc-ml-s"></oc-checkbox-stub></span></div>
        </th>
        <th class="oc-table-cell oc-table-cell-align-left oc-table-cell-align-middle oc-table-cell-width-shrink oc-text-nowrap oc-th oc-table-header-cell oc-table-header-cell-icon" style="top: 0px;">
          <div><span class="oc-table-thead-content header-text"></span></div>
        </th>
        <th class="oc-table-cell oc-table-cell-align-left oc-table-cell-align-middle oc-table-cell-width-auto oc-text-nowrap oc-th oc-table-header-cell oc-table-header-cell-name" style="top: 0px;" aria-sort="ascending"><button type="button" aria-label="Sort by name" class="oc-button oc-rounded oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-passive oc-button-passive-raw oc-button-sort oc-width-1-1">
            <!--v-if-->
            <!-- @slot Content of the button --> <span class="oc-table-thead-content header-text">Name</span> <span class="oc-icon oc-icon-s oc-icon-passive"><!----></span>
          </button></th>
        <th class="oc-table-cell oc-table-cell-align-left oc-table-cell-align-middle oc-table-cell-width-auto oc-text-nowrap oc-th oc-table-header-cell oc-table-header-cell-status" style="top: 0px;" aria-sort="none"><button type="button" aria-label="Sort by status" class="oc-button oc-rounded oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-passive oc-button-passive-raw oc-button-sort oc-width-1-1">
            <!--v-if-->
            <!-- @slot Content of the button --> <span class="oc-table-thead-content header-text">Status</span> <span class="oc-icon oc-icon-s oc-icon-passive oc-invisible-sr"><!----></span>
          </button></th>
        <th class="oc-table-cell oc-table-cell-align-left oc-table-cell-align-middle oc-table-cell-width-auto oc-text-nowrap oc-th oc-table-header-cell oc-table-header-cell-manager" style="top: 0px;">
          <div><span class="oc-table-thead-content header-text">Manager</span></div>
        </th>
        <th class="oc-table-cell oc-table-cell-align-left oc-table-cell-align-middle oc-table-cell-width-auto oc-text-nowrap oc-th oc-table-header-cell oc-table-header-cell-members" style="top: 0px;" aria-sort="none"><button type="button" aria-label="Sort by members" class="oc-button oc-rounded oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-passive oc-button-passive-raw oc-button-sort oc-width-1-1">
            <!--v-if-->
            <!-- @slot Content of the button --> <span class="oc-table-thead-content header-text">Members</span> <span class="oc-icon oc-icon-s oc-icon-passive oc-invisible-sr"><!----></span>
          </button></th>
        <th class="oc-table-cell oc-table-cell-align-left oc-table-cell-align-middle oc-table-cell-width-auto oc-text-nowrap oc-th oc-table-header-cell oc-table-header-cell-totalQuota" style="top: 0px;" aria-sort="none"><button type="button" aria-label="Sort by totalQuota" class="oc-button oc-rounded oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-passive oc-button-passive-raw oc-button-sort oc-width-1-1">
            <!--v-if-->
            <!-- @slot Content of the button --> <span class="oc-table-thead-content header-text">Total quota</span> <span class="oc-icon oc-icon-s oc-icon-passive oc-invisible-sr"><!----></span>
          </button></th>
        <th class="oc-table-cell oc-table-cell-align-left oc-table-cell-align-middle oc-table-cell-width-auto oc-text-nowrap oc-th oc-table-header-cell oc-table-header-cell-usedQuota" style="top: 0px;" aria-sort="none"><button type="button" aria-label="Sort by usedQuota" class="oc-button oc-rounded oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-passive oc-button-passive-raw oc-button-sort oc-width-1-1">
            <!--v-if-->
            <!-- @slot Content of the button --> <span class="oc-table-thead-content header-text">Used quota</span> <span class="oc-icon oc-icon-s oc-icon-passive oc-invisible-sr"><!----></span>
          </button></th>
        <th class="oc-table-cell oc-table-cell-align-left oc-table-cell-align-middle oc-table-cell-width-auto oc-text-nowrap oc-th oc-table-header-cell oc-table-header-cell-remainingQuota" style="top: 0px;" aria-sort="none"><button type="button" aria-label="Sort by remainingQuota" class="oc-button oc-rounded oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-passive oc-button-passive-raw oc-button-sort oc-width-1-1">
            <!--v-if-->
            <!-- @slot Content of the button --> <span class="oc-table-thead-content header-text">Remaining quota</span> <span class="oc-icon oc-icon-s oc-icon-passive oc-invisible-sr"><!----></span>
          </button></th>
        <th class="oc-table-cell oc-table-cell-align-left oc-table-cell-align-middle oc-table-cell-width-auto oc-text-nowrap oc-th oc-table-header-cell oc-table-header-cell-mdate" style="top: 0px;" aria-sort="none"><button type="button" aria-label="Sort by mdate" class="oc-button oc-rounded oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-passive oc-button-passive-raw oc-button-sort oc-width-1-1">
            <!--v-if-->
            <!-- @slot Content of the button --> <span class="oc-table-thead-content header-text">Modified</span> <span class="oc-icon oc-icon-s oc-icon-passive oc-invisible-sr"><!----></span>
          </button></th>
        <th class="oc-table-cell oc-table-cell-align-right oc-table-cell-align-middle oc-table-cell-width-auto oc-text-nowrap oc-th oc-table-header-cell oc-table-header-cell-actions oc-pr-s" style="top: 0px;">
          <div><span class="oc-table-thead-content header-text">Actions</span></div>
        </th>
      </tr>
    </thead>
    <tbody class="has-item-context-menu">
      <tr tabindex="-1" class="oc-tbody-tr oc-tbody-tr-1" data-item-id="1" draggable="false">
        <td class="oc-table-cell oc-table-cell-align-left oc-table-cell-align-middle oc-table-cell-width-shrink oc-td oc-table-data-cell oc-table-data-cell-select oc-pl-s">
          <oc-checkbox-stub id="oc-checkbox-2" disabled="false" modelvalue="false" option="undefined" label="Select 1 Some space" labelhidden="false" hidelabel="true" size="large" outline="false" class="oc-ml-s"></oc-checkbox-stub>
        </td>
        <td class="oc-table-cell oc-table-cell-align-left oc-table-cell-align-middle oc-table-cell-width-shrink oc-td oc-table-data-cell oc-table-data-cell-icon"><span class="oc-icon oc-icon-m oc-icon-passive"><!----></span></td>
        <td class="oc-table-cell oc-table-cell-align-left oc-table-cell-align-middle oc-table-cell-width-auto oc-td oc-table-data-cell oc-table-data-cell-name mark-element"><span class="spaces-table-space-name" data-test-space-name="1 Some space">1 Some space</span></td>
        <td class="oc-table-cell oc-table-cell-align-left oc-table-cell-align-middle oc-table-cell-width-auto oc-td oc-table-data-cell oc-table-data-cell-status"><span class="oc-flex oc-flex-middle"><span class="oc-icon oc-icon-s oc-icon-passive"><!----></span></span></td>
        <td class="oc-table-cell oc-table-cell-align-left oc-table-cell-align-middle oc-table-cell-width-auto oc-td oc-table-data-cell oc-table-data-cell-manager">user1, user2... +1</td>
        <td class="oc-table-cell oc-table-cell-align-left oc-table-cell-align-middle oc-table-cell-width-auto oc-td oc-table-data-cell oc-table-data-cell-members">3</td>
        <td class="oc-table-cell oc-table-cell-align-left oc-table-cell-align-middle oc-table-cell-width-auto oc-td oc-table-data-cell oc-table-data-cell-totalQuota">1 GB</td>
        <td class="oc-table-cell oc-table-cell-align-left oc-table-cell-align-middle oc-table-cell-width-auto oc-td oc-table-data-cell oc-table-data-cell-usedQuota">0 B</td>
        <td class="oc-table-cell oc-table-cell-align-left oc-table-cell-align-middle oc-table-cell-width-auto oc-td oc-table-data-cell oc-table-data-cell-remainingQuota">1 GB</td>
        <td class="oc-table-cell oc-table-cell-align-left oc-table-cell-align-middle oc-table-cell-width-auto oc-td oc-table-data-cell oc-table-data-cell-mdate"><span tabindex="0"></span></td>
        <td class="oc-table-cell oc-table-cell-align-right oc-table-cell-align-middle oc-table-cell-width-auto oc-td oc-table-data-cell oc-table-data-cell-actions oc-pr-s">
          <div class="spaces-list-actions"><button type="button" aria-label="Show details" class="oc-button oc-rounded oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-passive oc-button-passive-raw oc-mr-xs quick-action-button spaces-table-btn-details oc-p-xs">
              <!--v-if-->
              <!-- @slot Content of the button --> <span class="oc-icon oc-icon-m oc-icon-passive"><!----></span>
            </button> <button type="button" aria-label="Show context menu" class="oc-button oc-rounded oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-passive oc-button-passive-raw spaces-table-btn-action-dropdown" id="context-menu-trigger-1">
              <!--v-if-->
              <!-- @slot Content of the button --> <span class="oc-icon oc-icon-m oc-icon-passive"><!----></span>
              <div id="context-menu-drop-1" class="oc-drop oc-box-shadow-medium oc-rounded oc-overflow-hidden">
                <div class="oc-card oc-card-body oc-background-secondary oc-p-s"></div>
              </div>
            </button></div>
        </td>
      </tr>
      <tr tabindex="-1" class="oc-tbody-tr oc-tbody-tr-2" data-item-id="2" draggable="false">
        <td class="oc-table-cell oc-table-cell-align-left oc-table-cell-align-middle oc-table-cell-width-shrink oc-td oc-table-data-cell oc-table-data-cell-select oc-pl-s">
          <oc-checkbox-stub id="oc-checkbox-3" disabled="false" modelvalue="false" option="undefined" label="Select 2 Another space" labelhidden="false" hidelabel="true" size="large" outline="false" class="oc-ml-s"></oc-checkbox-stub>
        </td>
        <td class="oc-table-cell oc-table-cell-align-left oc-table-cell-align-middle oc-table-cell-width-shrink oc-td oc-table-data-cell oc-table-data-cell-icon"><span class="oc-icon oc-icon-m oc-icon-passive"><!----></span></td>
        <td class="oc-table-cell oc-table-cell-align-left oc-table-cell-align-middle oc-table-cell-width-auto oc-td oc-table-data-cell oc-table-data-cell-name mark-element"><span class="spaces-table-space-name" data-test-space-name="2 Another space">2 Another space</span></td>
        <td class="oc-table-cell oc-table-cell-align-left oc-table-cell-align-middle oc-table-cell-width-auto oc-td oc-table-data-cell oc-table-data-cell-status"><span class="oc-flex oc-flex-middle"><span class="oc-icon oc-icon-s oc-icon-passive"><!----></span></span></td>
        <td class="oc-table-cell oc-table-cell-align-left oc-table-cell-align-middle oc-table-cell-width-auto oc-td oc-table-data-cell oc-table-data-cell-manager">user1, user2... +1</td>
        <td class="oc-table-cell oc-table-cell-align-left oc-table-cell-align-middle oc-table-cell-width-auto oc-td oc-table-data-cell oc-table-data-cell-members">5</td>
        <td class="oc-table-cell oc-table-cell-align-left oc-table-cell-align-middle oc-table-cell-width-auto oc-td oc-table-data-cell oc-table-data-cell-totalQuota">2 GB</td>
        <td class="oc-table-cell oc-table-cell-align-left oc-table-cell-align-middle oc-table-cell-width-auto oc-td oc-table-data-cell oc-table-data-cell-usedQuota">500 MB</td>
        <td class="oc-table-cell oc-table-cell-align-left oc-table-cell-align-middle oc-table-cell-width-auto oc-td oc-table-data-cell oc-table-data-cell-remainingQuota">1.5 GB</td>
        <td class="oc-table-cell oc-table-cell-align-left oc-table-cell-align-middle oc-table-cell-width-auto oc-td oc-table-data-cell oc-table-data-cell-mdate"><span tabindex="0"></span></td>
        <td class="oc-table-cell oc-table-cell-align-right oc-table-cell-align-middle oc-table-cell-width-auto oc-td oc-table-data-cell oc-table-data-cell-actions oc-pr-s">
          <div class="spaces-list-actions"><button type="button" aria-label="Show details" class="oc-button oc-rounded oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-passive oc-button-passive-raw oc-mr-xs quick-action-button spaces-table-btn-details oc-p-xs">
              <!--v-if-->
              <!-- @slot Content of the button --> <span class="oc-icon oc-icon-m oc-icon-passive"><!----></span>
            </button> <button type="button" aria-label="Show context menu" class="oc-button oc-rounded oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-passive oc-button-passive-raw spaces-table-btn-action-dropdown" id="context-menu-trigger-2">
              <!--v-if-->
              <!-- @slot Content of the button --> <span class="oc-icon oc-icon-m oc-icon-passive"><!----></span>
              <div id="context-menu-drop-2" class="oc-drop oc-box-shadow-medium oc-rounded oc-overflow-hidden">
                <div class="oc-card oc-card-body oc-background-secondary oc-p-s"></div>
              </div>
            </button></div>
        </td>
      </tr>
    </tbody>
    <tfoot class="oc-table-footer">
      <tr class="oc-table-footer-row">
        <td colspan="11" class="oc-table-footer-cell">
          <!-- @slot Footer of the table -->
          <!--v-if-->
          <div class="oc-text-center oc-width-1-1 oc-my-s">
            <p class="oc-text-muted">2 spaces in total</p>
            <!--v-if-->
          </div>
        </td>
      </tr>
    </tfoot>
  </table>
</div>"
`;
